compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -728,12 +728,18 @@ class Typer extends Namer
   def typedIf(tree: untpd.If, pt: Type)(implicit ctx: Context): Tree = track("typedIf") {
     if (tree.isInline) checkInInlineContext("inline if", tree.posd)
     val cond1 = typed(tree.cond, defn.BooleanType)
-    val thenp2 :: elsep2 :: Nil = harmonic(harmonize, pt) {
-      val thenp1 = typed(tree.thenp, pt.notApplied)
-      val elsep1 = typed(tree.elsep.orElse(untpd.unitLiteral.withSpan(tree.span)), pt.notApplied)
-      thenp1 :: elsep1 :: Nil
+
+    if (tree.elsep `eq` untpd.EmptyTree) {
+      val thenp1 = typed(tree.thenp, defn.UnitType)
+      val elsep1 = tpd.unitLiteral.withSpan(tree.span)
```

Comment:
While you're at it, this should be:
```suggestion
      val elsep1 = tpd.unitLiteral.withSpan(tree.span.endPos)
```
Because it's the position of a synthetic tree that doesn't contain any user-written tree.